types/uint_test.go

```diff
@@ -171,6 +173,64 @@ func TestU64_Hex(t *testing.T) {
 	})
 }
 
+func TestUCompact_EncodeDecode(t *testing.T) {
+	bn := MustHexDecodeString("0x5C2D3BE75CEF559F050") //27205758526767196926032
+	uc := NewUCompact(big.NewInt(0).SetBytes(bn))
+
+	// Encode
+	var buffer = bytes.Buffer{}
+	err := scale.NewEncoder(&buffer).Encode(uc)
+	assert.NoError(t, err)
+	assert.Equal(t, buffer.Bytes(), MustHexDecodeString("0x1b50f059f5ce75bed3c205")) // Encoded number above
+
+	// Decode
+	dec := scale.NewDecoder(bytes.NewReader(buffer.Bytes()))
+	var res UCompact
+	err = dec.Decode(&res)
+	assert.NoError(t, err)
+	assert.Equal(t, uc, res)
+}
+
+func TestUCompact_EncodeDecode_MaxValue(t *testing.T) {
+	// Valid Max Number Encode/Decode [ 67 bytes -> MAX = (2**536)-1 ]
+	bigNumber := []byte{
+		0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+		0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+		0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+		0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+		0xff, 0xff, 0xff,
+	}
+	maxValue := NewUCompact(new(big.Int).SetBytes(bigNumber))
+	expectedEncoded := MustHexDecodeString("0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff")
+
+	var buffer = bytes.Buffer{}
+	err := scale.NewEncoder(&buffer).Encode(maxValue)
+	assert.NoError(t, err)
+	assert.Equal(t, buffer.Bytes(), expectedEncoded)
+
+	dec := scale.NewDecoder(bytes.NewReader(buffer.Bytes()))
+	var res UCompact
+	err = dec.Decode(&res)
+	assert.NoError(t, err)
+	assert.Equal(t, maxValue, res)
+
+	// Invalid max number Encode/Decode [ 68 bytes ]
+	bigNumber = append(bigNumber, []byte{0xff}...)
+	maxValue = NewUCompact(new(big.Int).SetBytes(bigNumber))
+
+	buffer = bytes.Buffer{}
+	err = scale.NewEncoder(&buffer).Encode(maxValue)
+	assert.Error(t, err)
+
+	// Decoding truncates at max length
+	expectedEncoded = append(expectedEncoded, []byte{0xab, 0xff, 0x34}...)
+	dec = scale.NewDecoder(bytes.NewReader(expectedEncoded))
+	var res1 UCompact
+	err = dec.Decode(&res1)
+	assert.NoError(t, err)
+	assert.Equal(t, res, res1)
```

Comment:
Very nice tests! Please add one more test case for negative `big.Int`, which should error. 

I also think that these tests should live in the scale codec (such that the codec package is self-cotained), maybe just copy the there?

Reply:
I will add the negative test case, but since this is in the types package and I want to compare the UCompact type, it would create a cyclic dep, so will keep it here for now.